File: ConstructBinaryTreefromPreorderandInorderTraversal.py
#!/usr/bin/env python
# -*- coding:utf-8 -*-   
# Author:zouzhipeng
"""
LeetCode 105
Given preorder and inorder traversal of a tree, construct the binary tree.

Note:
You may assume that duplicates do not exist in the tree.
"""


# Definition for a binary tree node.
class TreeNode(object):
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution(object):
    def buildTree(self, preorder, inorder):
        """
        :type preorder: List[int]
        :type inorder: List[int]
        :rtype: TreeNode
        """
        if len(preorder) == 0 or len(preorder) != len(inorder):
            return None
        root = TreeNode(preorder[0])
        if len(preorder) == 1:
            return root
        for i, key in enumerate(inorder):
            if key == preorder[0]:
                index = i
                break
        left = self.buildTree(preorder[1:index+1], inorder[:index])
        right = self.buildTree(preorder[index+1:], inorder[index+1:])
        root.left = left
        root.right = right
        return root

"""
leetcode 106
Construct Binary Tree from Inorder and Postorder Traversal
中序遍历和后续遍历，求二叉树
"""


class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        if len(postorder) > 0:
            index = inorder.index(postorder[-1])
            root = TreeNode(inorder[index])
            root.left = self.buildTree(inorder[:index], postorder[:index])
            root.right = self.buildTree(inorder[index+1:], postorder[index:-1])
            return root
        return None

File: c++/ConstructBinaryTreefromPreorderandInorderTraversal.cpp
#include <iostream>
#include <vctor>
using namespace std;
using std::vector;

//Definition for a binary tree node.
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};


class Solution {
public:
    TreeNode* buildRoot(vector<int>& preorder, vector<int>& inorder, int i , int j, int k, int l) {
        if(j<i) return nullptr;
        TreeNode* node = new TreeNode(preorder[i]);
        int pos;
        for(int t = k;t<=l;t++) {
            if(inorder[t] == preorder[i]) {
                pos = t;
                break;
            }
        }
        node->left = buildRoot(preorder, inorder, i+1, i+pos-k, k, pos-1);
        node->right = buildRoot(preorder, inorder, j-k+pos+1, j,pos+1, l);
        return node;
    }

    TreeNode* buildTree(vector<int>& preorder, vector<int>& inorder) {
        return buildRoot(preorder, inorder, 0, preorder.size()-1, 0, inorder.size()-1);
    }
};


// 已知中序和后序，求二叉树
class Solution {
public:
    TreeNode* help(vector<int>& inorder, vector<int>& postorder, int is, int ie, int ps, int pe) {
        if(is > ie) return nullptr;
        TreeNode* node = new TreeNode(postorder[pe]);
        int pos;
        for(int i = is;i<=ie;i++) {
            if(inorder[i] == postorder[pe]) {
                pos = i;
                break;
            }
        }
        node->left = help(inorder, postorder, is, pos-1, ps, ps+pos-is-1);
        node->right = help(inorder, postorder, pos+1, ie, ps+pos-is, pe-1);
        return node;
    }

    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return help(inorder, postorder, 0, inorder.size()-1, 0, postorder.size()-1);
    }
};

// 1 2 3 4 5
// 4 3 5 2 1